### PR TITLE
Add PoW engine

### PR DESCRIPTION
**Pull request by WTer**
Interface and implementation.

### CODE REVIEW

File: consensus/pow/sha256worker.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    "crypto/sha256"
+    "math"
+    "math/big"
+)
+
+// data isathe pointer to the target block data,
+// target is the goal, which means we need to mine a result and its hash must be less than the target.
+type SHA256Worker struct {
+    Data   []byte
+    Nonce  int
```
Comment (lucusfly) [inline consensus/pow/sha256worker.go]: nonce may be bigger than int.Max

File: consensus/pow/sha256worker.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    return data
+}
+
+// Loop nonce to find the target value that meet the requirement.
+func (worker *SHA256Worker) Producer() int {
+    nonce := 0
+    var hash [32]byte
+    var hashInt big.Int
+
+    for nonce < math.MaxInt64 {
```
Comment (lucusfly) [inline consensus/pow/sha256worker.go]: math.MaxInt64?

File: consensus/pow/utils.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package consensus
+
```
Comment (lucusfly) [inline consensus/pow/utils.go]: move to common folder

File: consensus/pow/utils.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import (
+    "bytes"
+    "encoding/binary"
+    "log"
+)
+
+// Converts a int64 value into a byte array
+func Int2Hex(num int64) []byte {
```
Comment (lucusfly) [inline consensus/pow/utils.go]: Int2Hex should be Int2Bytes

File: consensus/pow/worker.go [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package consensus
+
+// Worker is an PoW engine.
+type Worker interface {
+    // Returns the current mining result of a PoW consensus engine.
+    Producer() int
```
Comment (lucusfly) [inline consensus/pow/worker.go]: Producer() int
better to return a string if you want it to be general

File: consensus/pow/worker.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package consensus
+
```
Comment (lucusfly) [inline consensus/pow/worker.go]: move up to consensus folder for interface

File: consensus/pow/sha256worker.go (hunk 1 above)
Comment (WTer) [reply consensus/pow/sha256worker.go]: Good catch, fixed it.

File: consensus/pow/sha256worker.go (hunk 2 above)
Comment (WTer) [reply consensus/pow/sha256worker.go]: I changed the definition of nonce from int to int64.

File: consensus/pow/utils.go (hunk 3 above)
Comment (WTer) [reply consensus/pow/utils.go]: consensus is a independent component, there is also consensus folder in ETH. so I would prefer to keep it.

File: consensus/pow/utils.go (hunk 4 above)
Comment (WTer) [reply consensus/pow/utils.go]: nice.

File: consensus/pow/worker.go (hunk 5 above)
Comment (WTer) [reply consensus/pow/worker.go]: nice suggestion. fixed.

File: consensus/pow/worker.go (hunk 6 above)
Comment (WTer) [reply consensus/pow/worker.go]: just keep it.

File: consensus/pow/sha256worker.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    "bytes"
+    "crypto/sha256"
+    "math"
+    "math/big"
+)
+
+// data isathe pointer to the target block data,
+// target is the goal, which means we need to mine a result and its hash must be less than the target.
+type SHA256Worker struct {
+    Data   []byte
```
Comment (boqiu) [inline consensus/pow/sha256worker.go]: Comment is required for all exported type/variable, and better to follow the golint  guideline.

Comment (WTer) [reply consensus/pow/sha256worker.go]: nice.

File: common/converter.go [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package common
+
+import (
+    "bytes"
+    "encoding/binary"
+    "log"
+)
+
+// Int2Bytes Converts a int64 value into a byte array
+func Int2Bytes(num int64) []byte {
```
Comment (boqiu) [inline common/converter.go]: Confused func name, what about ConvertInt64ToBytes?

File: consensus/pow/sha256/sha256worker.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
+package sha256
```
Comment (boqiu) [inline consensus/pow/sha256/sha256worker.go]: How about change file name to worker.go

File: common/converter.go (hunk 8 above)
Comment (WTer) [reply common/converter.go]: sure, changed.

File: consensus/pow/sha256/sha256worker.go (hunk 9 above)
Comment (WTer) [reply consensus/pow/sha256/sha256worker.go]: nice suggestion.

File: consensus/pow/worker.go [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package pow
+
+// Worker is an PoW engine.
+type Worker interface {
+    // Returns the current mining result of a PoW consensus engine.
+    Producer() string
+
+    // Verify whether the mining result is meet the requirement.
+    Validator() bool
```
Comment (boqiu) [inline consensus/pow/worker.go]: Validate

File: consensus/pow/worker.go [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package pow
+
+// Worker is an PoW engine.
+type Worker interface {
+    // Returns the current mining result of a PoW consensus engine.
+    Producer() string
+
+    // Verify whether the mining result is meet the requirement.
```
Comment (boqiu) [inline consensus/pow/worker.go]: redundant "is"

File: consensus/pow/worker.go (hunk 10 above)
Comment (WTer) [reply consensus/pow/worker.go]: changed.

File: consensus/pow/worker.go (hunk 11 above)
Comment (WTer) [reply consensus/pow/worker.go]: removed.

**Pull request merged**